File: app/src/main/java/com/example/optiweather/repository/WeatherRepository.java
package com.example.optiweather.repository;

import androidx.annotation.NonNull;

import com.example.optiweather.model.WeatherData;
import com.example.optiweather.network.RetrofitClient;
import com.example.optiweather.network.WeatherApi;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class WeatherRepository {

    private final WeatherApi api;

    public WeatherRepository() {
        api = RetrofitClient.getApi();
    }

    public void getWeather(double lat, double lon, final WeatherCallback callback) {
        Call<WeatherData> call = api.getCurrentWeather(lat, lon, true);
        call.enqueue(new Callback<>() {
            @Override
            public void onResponse(@NonNull Call<WeatherData> call, @NonNull Response<WeatherData> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body());
                } else {
                    callback.onError("Response error: " + response.code());
                }
            }

            @Override
            public void onFailure(@NonNull Call<WeatherData> call, @NonNull Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }

    public interface WeatherCallback {
        void onSuccess(WeatherData weatherData);
        void onError(String error);
    }
}

File: app/src/main/java/com/example/optiweather/repository/WeatherRepository.kt
package com.example.optiweather.repository

import com.example.optiweather.model.WeatherData
import com.example.optiweather.network.RetrofitClient
import com.example.optiweather.network.WeatherApi
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class WeatherRepository {
    private val api: WeatherApi = RetrofitClient.api

    fun getWeather(lat: Double, lon: Double, callback: WeatherCallback) {
        val call = api.getCurrentWeather(lat, lon, true)
        call!!.enqueue(object : Callback<WeatherData?> {
            override fun onResponse(call: Call<WeatherData?>, response: Response<WeatherData?>) {
                if (response.isSuccessful && response.body() != null) {
                    callback.onSuccess(response.body())
                } else {
                    callback.onError("Response error: " + response.code())
                }
            }

            override fun onFailure(call: Call<WeatherData?>, t: Throwable) {
                callback.onError(t.message)
            }
        })
    }

    interface WeatherCallback {
        fun onSuccess(weatherData: WeatherData?)
        fun onError(error: String?)
    }
}